internal/ppa/ppa.go

```diff
@@ -0,0 +1,377 @@
+		return nil
+	}
+
+	ppaFile := filepath.Join(sourcesListD, fileName)
+	err := osRemove(ppaFile)
+	if err != nil {
+		return fmt.Errorf("Error removing %s: %s", ppaFile, err.Error())
+	}
+	return nil
+}
+
+func (p *BasePPA) importKey(basePath string, ppaFileName string, ascii bool, debug bool) (err error) {
+	trustedGPGD := filepath.Join(basePath, trustedGPGDPath)
+	keyFileName := strings.Replace(ppaFileName, ".list", ".gpg", 1)
+	keyFilePath := filepath.Join(trustedGPGD, keyFileName)
+
+	err = p.ensureFingerprint(lpBaseURL)
+	if err != nil {
+		return err
+	}
+
+	tmpGPGDir, err := p.createTmpGPGDir(basePath)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		tmpErr := osRemoveAll(tmpGPGDir)
+		if tmpErr != nil {
+			if err != nil {
+				err = fmt.Errorf("%s after previous error: %w", tmpErr.Error(), err)
+			} else {
+				err = fmt.Errorf("Error removing temporary gpg directory \"%s\": %s", tmpGPGDir, tmpErr.Error())
+			}
+		}
+	}()
+
+	tmpASCIIKeyFilName := keyFileName + ".asc"
+	tmpASCIIKeyFilePath := filepath.Join(tmpGPGDir, tmpASCIIKeyFilName)
+
+	commonGPGArgs := []string{
+		"--no-default-keyring",
+		"--no-options",
+		"--batch",
+		"--homedir",
+		tmpGPGDir,
+		"--secret-keyring",
+		filepath.Join(tmpGPGDir, "tempring.gpg"),
+		"--keyserver",
+		"hkp://keyserver.ubuntu.com:80",
+	}
+	recvKeyArgs := append(commonGPGArgs, "--recv-keys", p.Fingerprint)
+
+	exportKeyArgs := make([]string, 0)
+	exportKeyArgs = append(exportKeyArgs, commonGPGArgs...)
+
+	if ascii {
+		exportKeyArgs = append(exportKeyArgs, "-a", "--output", tmpASCIIKeyFilePath)
+	} else {
+		exportKeyArgs = append(exportKeyArgs, "--output", keyFilePath)
+	}
+
+	exportKeyArgs = append(exportKeyArgs, "--export", p.Fingerprint)
+
+	gpgCmds := []*exec.Cmd{
+		execCommand(
+			"gpg",
+			recvKeyArgs...,
+		),
+		execCommand(
+			"gpg",
+			exportKeyArgs...,
+		),
+	}
+
+	for _, gpgCmd := range gpgCmds {
+		gpgOutput := helper.SetCommandOutput(gpgCmd, debug)
+		err := gpgCmd.Run()
+		if err != nil {
+			err = fmt.Errorf("Error running gpg command \"%s\". Error is \"%s\". Full output below:\n%s",
+				gpgCmd.String(), err.Error(), gpgOutput.String())
+			return err
+		}
+	}
+
+	keyBytes := []byte{}
+	if ascii {
+		keyBytes, err = os.ReadFile(tmpASCIIKeyFilePath)
+		if err != nil {
+			return err
+		}
+	}
+
+	p.signingKey = string(keyBytes)
+
+	return nil
+}
+
+// ensureFingerprint ensures a non empty fingerprint is set on the PPA object
+func (p *BasePPA) ensureFingerprint(baseURL string) error {
+	if p.PPA.Fingerprint != "" {
+		return nil
+	}
+	// The YAML schema has already been validated that if no fingerprint is
+	// provided, then this is a public PPA. We will get the fingerprint
+	// from the Launchpad API
+	type lpResponse struct {
+		SigningKeyFingerprint string `json:"signing_key_fingerprint"`
+		// plus many other fields that aren't needed at the moment
+	}
+	lpRespContent := &lpResponse{}
+
+	lpURL := fmt.Sprintf("%s/devel/~%s/+archive/ubuntu/%s", baseURL,
```

Comment:
So yeah just fetch the key directly by appending `?ws.op=getSigningKeyData` and then unmarshalling the result directly into a string or `type lpResponse string` if you like :)

Reply:
It seems this end point also does not work for private PPAs, the fingerprint you get from it is redacted:

```
$ env curl -s https://api.launchpad.net/devel/~ubuntu-uefi-team/+archive/ubuntu/proposed | jq  | grep signing
  "signing_key_fingerprint": "tag:launchpad.net:2008:redacted",
```